### DISCUSSION TITLE
Is there any documentation about JSI?

### DISCUSSION THREAD

**Discussion by samrith-s** [reactwg/react-native-new-architecture#48] [Q&A] [answered]
I am exploring moving to the new architecture at our company, and was wondering if we have any documentation about JSI and the React Native Core someplace.

We have developers who are comfortable with C++, Java/Kotlin, Objective-C/Swift and would love to move some of our stuff over to the new architecture but are finding it difficult to grasp as there is no documentation surrounding it.

**Comment by cipolleschi** [comment, marked as answer]
Hi @samrith-s, thanks for your question!

> I am exploring moving to the new architecture at our company, and was wondering if we have any documentation about JSI and the React Native Core someplace.

What kind of documentation would you like to have? We were not planning to have it, given that it is some internal details and you users won't have to deal with JSI directly. But if you help us out, understanding which kind of information you would like to find there, we can think about writing it and plan some work accordingly.

In the mean time, the source code for JSI is open source, and can be explored [here](https://github.com/facebook/react-native/tree/main/ReactCommon/jsi).

**Reply by samrith-s** [reply]
Hello @cipolleschi,

I was hoping for something like a detailed API documentation (à la JavaDoc)so that if we want to use it internally to build modules using JSI, we can easily refer to that. Right now it is a little bit cumbersome, and a web page which lets us explore APIs would be amazing and pretty helpful to the community as a whole I believe.

**Reply by sammy-SC** [reply]
Hey @samrith-s,

may I ask why not build a native module? Have you ran into any limitations with native modules? Using native modules is preferred way of calling native code from JavaScript.

Unfortunately like @cipolleschi mentioned, there isn't dedicated documentation for JSI. If you really want to use JSI, I recommend looking at [UIManagerBinding.cpp](https://github.com/facebook/react-native/blob/main/ReactCommon/react/renderer/uimanager/UIManagerBinding.cpp) or [RuntimeSchedulerBinding.cpp](https://github.com/facebook/react-native/blob/main/ReactCommon/react/renderer/runtimescheduler/RuntimeSchedulerBinding.cpp). I think these two classes show well how to use JSI to define functions on the runtime.

**Reply by samrith-s** [reply]
Hello @sammy-SC 

We do use native modules. But what we wanted was to develop a countdown timer module which ticks down to 0 given the start time. And we were thinking of keeping the base implementation in C++ (using JSI) so that our core logic is shared across platforms and we expose other platform-specific stuff via Java/Kotlin/Objective-C/Swift bindings.

Thanks for the tip, will go through the two files and gather whatever I can!

**Reply by cipolleschi** [reply]
Hi @samrith-s, out of curiosity, have you tried to develop a TurboModule using a pure C++ implementation? This allows you to:
- Use JSI without using it directly, as @sammy-SC was suggesting
- Share code between all the platforms you need

**Reply by teodorciuraru** [reply]
@cipolleschi I tried. Whatever method you create in C++, you must redefine it (marshal inside) inside the platforms—at least with the current create-react-native-library template, resulting in duplication. I switched to implementing native module + JSI which has C++ as the only source of truth, but I had to reverse engineer a lot of stuff because documentation was lacking.

**Reply by cipolleschi** [reply]
Have you had a look at the [documentation](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/turbo-modules-xplat.md) on how to create a pure C++ turbomodule? Would this be a good solution for the problem?

**Reply by teodorciuraru** [reply]
@cipolleschi Yep. It looks more like _how to add pure C++ TurboModule to an RN app_, not **how to create a Turbo Module library** with only a C++ source of truth – that can be used inside other projects. I'm not sure if it's expected to create an RN library project only from that sample. I, myself, find it non-trivial to do. What I find missing is a `create-react-native-library` TurboModules with **pure** C++ (no platform marshaling).

**Reply by cortinico** [reply]
@teodorciuraru here a sample repository with a standalone C++ TurboModule that works with autolinking (at least on Android). Adding iOS support should be that hard https://github.com/cortinico/reproducer-cxx-tm-autolinking
